### PR TITLE
Set permissions on log directory

### PR DESCRIPTION
**Pull request by Jontified**
Move a bit of the C++ code to rust and make sure that log directory is protected with admin permissions.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)

### CODE REVIEW

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 7 of 14 files at r1, 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: 9 of 14 files reviewed, 2 unresolved discussions (waiting on @Jontified)

___
*[`mullvad-paths/src/windows.rs` line 71 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NiaihTpAtTYRO21bVHw:-NiaihTpAtTYRO21bVHx:bgolzs9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L71)):*
> ```XML
> /// only directories that do not already exist and the leaf directory will have their permissions set.
> pub fn create_dir_recursive(path: &Path, set_security_permissions: bool) -> Result<()> {
>     if set_security_permissions {
> ```

Are the branches flipped? `set_security_permissions == true` seems to imply that permissions should be set.
___
*[`windows/windows-libraries` line 1 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Niag-t48WMzCF9nPSEo:-Niag-t48WMzCF9nPSEp:b-buobfy) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/windows/windows-libraries#L1)):*
> ```
> Subproject commit 6df03881c5bbde740ba601906594be61b99319a7
> ```

In my opinion, wrapping `create_privileged_directory` in `libcommon` creates a strange dependency for `libcommon` on the app repository. That's supposed to be an independent library. Moreover, the point of `mullvad-nsis` it to depend less on it, if anything.

I suggest either (a) use `create_privileged_directory` directly in the projects, or (b) put the wrapper in `libshared` (not a submodule).

**Review by dlon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: 9 of 14 files reviewed, 3 unresolved discussions (waiting on @Jontified)

___
*[`mullvad-paths/src/windows.rs` line 146 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Nial0An-t1dlOAzZ1lC:-Nial0An-t1dlOAzZ1lD:b-w0l8dg) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L146)):*
> ```XML
>     // Authenticated users SID https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/manage/understand-security-identifiers
>     let (authenticated_users_ea, authenticated_users_psid) =
>         match create_explicit_access("S-1-5-11", GENERIC_READ) {
> ```

This is fine, but it seems like we could have used `CreateWellKnownSid(WinAuthenticatedUserSid, ...)` and `CreateWellKnownSid(WinBuiltinAdministratorsSid, ...)` here.

A nice thing about it is that we allocate the memory ourselves, so most of the `LocalFree`s go away, and we only have to use plain arrays or `Vec`s. Cleaner, probably?

Though I vaguely recall that this did not work for some reason.
___
*[`mullvad-paths/src/windows.rs` line 217 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NiajRej5ug9d_p-Xs--:-NiajRej5ug9d_p-Xs-0:b-237n85) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L217)):*
> ```XML
>     let sid_wide_str = get_wide_str(sid_wide_str);
>     let mut psid = ptr::null_mut();
>     if 0 == unsafe { ConvertStringSidToSidW(sid_wide_str.as_ptr(), &mut psid) } {
> ```

I think we should avoid Yoda conditions, as they're inconsistent with our general code style. The point of these is mainly to prevent accidental assignment, which isn't relevant for if statements in Rust.

File: mullvad-paths/src/windows.rs [hunk 1]
```diff
@@ -41,6 +56,189 @@ impl Drop for Handle {
...
+fn get_wide_str<S: AsRef<OsStr>>(string: S) -> Vec<u16> {
+    let wide_string: Vec<u16> = string.as_ref()
+        .encode_wide()
+        // Add null terminator
+        .chain(std::iter::once(0))
+        .collect();
+    wide_string
+}
+
+/// Recursively creates directories for the given path with the given security attributes
```
Comment (faern) [inline mullvad-paths/src/windows.rs:68]: > with the given security attributes

This function does not take any security attributes.

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: 8 of 14 files reviewed, 4 unresolved discussions (waiting on @dlon and @faern)

___
*[`mullvad-paths/src/windows.rs` line 68 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#gh-1384547173:-NidFR_yCbV5U8oUPabL:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L68)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

> with the given security attributes

This function does not take any security attributes.
</blockquote></details>

Done.
___
*[`mullvad-paths/src/windows.rs` line 71 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NiaihTpAtTYRO21bVHw:-NichXyRBUv4yT4_sbUf:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L71)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Are the branches flipped? `set_security_permissions == true` seems to imply that permissions should be set.
</blockquote></details>

Done.
___
*[`mullvad-paths/src/windows.rs` line 217 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NiajRej5ug9d_p-Xs--:-Nici6Su6ze6nbScvovf:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/mullvad-paths/src/windows.rs#L217)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

I think we should avoid Yoda conditions, as they're inconsistent with our general code style. The point of these is mainly to prevent accidental assignment, which isn't relevant for if statements in Rust.
</blockquote></details>

Done.
___
*[`windows/windows-libraries` line 1 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Niag-t48WMzCF9nPSEo:-NifKv1TDZ_BNxnv_a5-:bw0xxwc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/windows/windows-libraries#L1)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

In my opinion, wrapping `create_privileged_directory` in `libcommon` creates a strange dependency for `libcommon` on the app repository. That's supposed to be an independent library. Moreover, the point of `mullvad-nsis` it to depend less on it, if anything.

I suggest either (a) use `create_privileged_directory` directly in the projects, or (b) put the wrapper in `libshared` (not a submodule).
</blockquote></details>

Done. Since it's only called in `log` I think it's cleaner to just have the call there.

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 2 of 14 files at r1, 1 of 1 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: all files reviewed, 3 unresolved discussions (waiting on @faern and @Jontified)

___
*[`mullvad-nsis/Cargo.toml` line 18 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NifwGq-2Oqsd2z4MHP3:-NifwGq-2Oqsd2z4MHP4:b-gu0v0j) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/mullvad-nsis/Cargo.toml#L18)):*
> ```TOML
> 
> [target.i686-pc-windows-msvc.build-dependencies]
> cbindgen = { version = "0.24.3", default-features = false }
> ```

Nit: Diff just removes a newline.
___
*[`mullvad-paths/src/windows.rs` line 245 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Nifw5XVBAEpXqjL7mTS:-Nifw5XVBAEpXqjL7mTT:bpkuyo5) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/mullvad-paths/src/windows.rs#L245)):*
> ```XML
>     unsafe { LocalFree(new_dacl as isize) };
> 
>     if ERROR_SUCCESS != result {
> ```

Still one over here, but it's not a big deal.
___
*[`windows/nsis-plugins/src/log/log.cpp` line 301 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Nify4796w8YuSdvoPY6:-NifwOmiDYZ_jLL3Y3gQ:b-8impbz) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/windows/nsis-plugins/src/log/log.cpp#L301)):*
> ```cpp
>         if (Status::Ok != create_privileged_directory(reinterpret_cast<const uint16_t*>(w_path)))
>         {
>             THROW_ERROR("Failed to create privileged directory");
> ```

Nit: Fine, but why a needlessly vague "privileged directory" rather than "log directory"?
___
*[`windows/nsis-plugins/src/log/log.vcxproj` line 103 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Nifxlr24QAWhU3y5Skd:-NifxC6PBd7Ed1q0b_LU:brmwmv0) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/windows/nsis-plugins/src/log/log.vcxproj#L103)):*
> ```XML
>       <ModuleDefinitionFile>log.def</ModuleDefinitionFile>
>     </Link>
>   </ItemDefinitionGroup>
> ```

Looks like this file is missing a pre build event (see `cleanup.vcxproj`).
___
*[`windows/nsis-plugins/src/tray/tray.vcxproj` line 72 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NifwjEf0jFPtNSGgehz:-NifwjEg7bbG9CxuzI6k:bxrl5m8) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/windows/nsis-plugins/src/tray/tray.vcxproj#L72)):*
> ```XML
>       <GenerateDebugInformation>true</GenerateDebugInformation>
>       <ImageHasSafeExceptionHandlers>false</ImageHasSafeExceptionHandlers>
>       <AdditionalLibraryDirectories>$(ProjectDir)../../../../target/i686-pc-windows-msvc/release;$(ProjectDir)../../../../dist-assets/binaries/x86_64-pc-windows-msvc/nsis/;$(SolutionDir)bin\$(Platform)-$(Configuration)\</AdditionalLibraryDirectories>
> ```

Can the changes be reverted for this project?
___
*[`windows/windows-libraries` line 1 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Niag-t48WMzCF9nPSEo:-Niaj9_o5-vNY5j1VyxP:bccpe83) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/6c3b5cdc97454da9efdca230d1a21d4bea28fb1b/windows/windows-libraries#L1)):*
<details><summary><i>Previously, Jontified wrote…</i></summary><blockquote>

Done. Since it's only called in `log` I think it's cleaner to just have the call there.
</blockquote></details>

Awesome!

**Review by dlon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: all files reviewed, 4 unresolved discussions (waiting on @faern and @Jontified)

___
*[`windows/nsis-plugins/src/log/log.cpp` line 297 at r6](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-Nig-WPaCg_HDfnXYbzt:-Nig-WPb-XeoE1Lwqi2j:b-b1dh9h) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d52bfe705fe847ff022cfb8ade5390bb5575258a/windows/nsis-plugins/src/log/log.cpp#L297)):*
> ```cpp
> 		logpath.append(L"Mullvad VPN");
> 
>         const wchar_t* w_path = logpath.wstring().c_str();
> ```

Should be tabs.

**Review by dlon — APPROVED** [review]

**Review by dlon — APPROVED** [review]
> Reviewed 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: all files reviewed, 1 unresolved discussion (waiting on @faern)

**Review by dlon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: all files reviewed, 1 unresolved discussion (waiting on @Jontified)

___
*[`windows/nsis-plugins/src/log/log.cpp` line 297 at r7](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NikrBkp-vAnbaOednpc:-NikrBkp-vAnbaOednpd:bvlqd55) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/21714696fa50460811fefde7d840dadb540edd9e/windows/nsis-plugins/src/log/log.cpp#L297)):*
> ```cpp
> 		logpath.append(L"Mullvad VPN");
> 
> 		const wchar_t* w_path = logpath.wstring().c_str();
> ```

Use after free -- the `std::wstring` is destroyed here.

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: 13 of 14 files reviewed, 1 unresolved discussion (waiting on @dlon)

___
*[`windows/nsis-plugins/src/log/log.cpp` line 297 at r7](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398#-NikrBkp-vAnbaOednpc:-Nj6U1DN0lGfCgXGkoYI:b-4c5ezp) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/21714696fa50460811fefde7d840dadb540edd9e/windows/nsis-plugins/src/log/log.cpp#L297)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Use after free -- the `std::wstring` is destroyed here.
</blockquote></details>

Good catch, done.

**Review by dlon — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/5398)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**